src/Workspaces/Core/Portable/Utilities/TaskQueue.cs

```diff
@@ -0,0 +1,102 @@
+        public Task ScheduleTask(string taskName, Action operation, CancellationToken cancellationToken)
+            => EndOperation(BeginOperation(taskName), ScheduleTaskInProgress(operation, cancellationToken));
+
+        public Task<T> ScheduleTask<T>(string taskName, Func<T> operation, CancellationToken cancellationToken)
+            => EndOperation(BeginOperation(taskName), ScheduleTaskInProgress(operation, cancellationToken));
+
+        public Task ScheduleTask(string taskName, Func<Task> operation, CancellationToken cancellationToken)
+            => EndOperation(BeginOperation(taskName), ScheduleTaskInProgress(operation, cancellationToken));
+
+        public Task<T> ScheduleTask<T>(string taskName, Func<Task<T>> operation, CancellationToken cancellationToken)
+            => EndOperation(BeginOperation(taskName), ScheduleTaskInProgress(operation, cancellationToken));
+
+        [PerformanceSensitive("https://developercommunity.visualstudio.com/content/problem/854696/changing-target-framework-takes-10-minutes-with-10.html", AllowCaptures = false)]
+        public Task ScheduleTaskInProgress(Action operation, CancellationToken cancellationToken)
+        {
+            lock (_gate)
+            {
+                var task = _latestTask.SafeContinueWith(_ => operation(), cancellationToken, TaskContinuationOptions.None, _taskScheduler);
+                _latestTask = task;
+                return task;
+            }
+        }
+
+        [PerformanceSensitive("https://developercommunity.visualstudio.com/content/problem/854696/changing-target-framework-takes-10-minutes-with-10.html", AllowCaptures = false)]
+        public Task<T> ScheduleTaskInProgress<T>(Func<T> operation, CancellationToken cancellationToken)
+        {
+            lock (_gate)
+            {
+                var task = _latestTask.SafeContinueWith(_ => operation(), cancellationToken, TaskContinuationOptions.None, _taskScheduler);
+                _latestTask = task;
+                return task;
+            }
+        }
+
+        [PerformanceSensitive("https://developercommunity.visualstudio.com/content/problem/854696/changing-target-framework-takes-10-minutes-with-10.html", AllowCaptures = false)]
+        public Task ScheduleTaskInProgress(Func<Task> operation, CancellationToken cancellationToken)
+        {
+            lock (_gate)
+            {
+                var task = _latestTask.SafeContinueWithFromAsync(_ => operation(), cancellationToken, TaskContinuationOptions.None, _taskScheduler);
+                _latestTask = task;
+                return task;
+            }
+        }
+
+        [PerformanceSensitive("https://developercommunity.visualstudio.com/content/problem/854696/changing-target-framework-takes-10-minutes-with-10.html", AllowCaptures = false)]
+        public Task<T> ScheduleTaskInProgress<T>(Func<Task<T>> operation, CancellationToken cancellationToken)
```

Comment:
as before, i don't really understand the InProgress moniker here.  can you explain what you want ot convey?  then the name may make sense, or we an come up with other name.

Reply:
I couldn't come up with a better one, feel free to suggest some. What it means is that this methods does not "start" nor "end" the task (in terms of begin/end async operation). It queues the operation as a part of ongoing (in-progress) work.

Reply:
ah, i see.  Could you jsut doc the methods with that information.  Note: i'm surprised we don't still start/end things with this.  Should we?  it seems weird that we'd have a task in the list running, but hte overal listener would say "i'm done".

Why not start, then call the func to create the task.  then have that task still end things?   I get that you'll get back a hot-task.  but it still seems like it *should* work with Begin/End.

Reply:
This is only used in one scenario. The caller calls the listener BeginAsyncOperation then it calls ScheduleTaskInProgress and then it calls Completes. So, the sequence of calls is the same as ScheduleTask but it's not a single method call - it's spread out. If you search the PR for ScheduleTaskInProgress  you'll find the pattern.